bson.c

```diff
@@ -849,25 +849,43 @@ char* bson_to_zval(char *buf, HashTable *result, mongo_bson_conversion_options *
 	 * bytes to that point.
 	 *
 	 * We lose buf's position as we iterate, so we need buf_start to save it. */
-	char *buf_start = buf, *buf_end;
+	const char *buf_start = buf, *buf_end;
 	unsigned char type;
 
 	if (buf == 0) {
 		return 0;
 	}
 
-	buf_end = buf + MONGO_32(*((int*)buf));
+	if (buf_len < 5) {
+		zend_throw_exception_ex(mongo_ce_CursorException, 38 TSRMLS_CC, "BSON buffer is %u bytes, but should be >= 5 (i.e. empty document)", buf_len);
+		return 0;
+	}
+
+	if (buf_len < (size_t) MONGO_32(*((int32_t*)buf))) {
+		zend_throw_exception_ex(mongo_ce_CursorException, 39 TSRMLS_CC, "BSON buffer is %u bytes, but document length is %u", buf_len, MONGO_32(*((int32_t*)buf)));
+		return 0;
```

Comment:
In the case for `BSON_ARRAY` and `BSON_OBJECT` (I'll refer to both as an embedded document), we peek at the first int32 to capture the reported length and store that in `doc_len`. We then compare that against the current document's length when calling `CHECK_BUFFER_LEN(doc_len)`:

```
if (buf + (len) >= buf_end) {
    /* throw */
}
```

If the embedded document pretends to exceed the span of its parent document (be it the top-level doc or another embedded doc), we'll throw an exception before recursing into another `bson_to_zval()` call. I'd wager that in a recursive call, this second overflow check will always pass. I'm not sure it's worth passing an extra boolean argument like `is_buf_len_trustworthy`, though. We can certainly benchmark that if you like.

That said, I will add some additional tests for embedded document overflows (you mentioned we were testing any nested error cases).